UH-StudentServices/student_guide
# Feature/hub 219 teacher office times

# Displaying office hours on Guide

Teachers enter and maintain their office hours in My Teaching service (teacher.helsinki.fi). Those hours will be displayed also in Guide. This pull request implements the integration for getting the office hours and displaying them in Guide.

The office hours can be displayed in any instruction (article) by adding an instance of the new "Office hours" paragraph. The paragraph itself can be positioned the same way as other paragraphs, but contains no editable content. The display logic is handled by the paragraph (no need for content editors to do anything else in addition to adding the paragraph).

When no degree programme is selected by the user, all office hours are displayed. When a degree programme is selected, all office hours with the same degree programme are displayed (others are hidden). The displaying/hiding is handled in the client (ECMAScript) to keep things simple (it is simple to cache the whole office hours response and let the filtering happen in the UI layer).

The office hours are cached for 15 minutes to avoid unnecessary requests. Only successful responses are cached. Other than successful responses are logged for troubleshooting.

## Testing notes

- Run provisioning (there are settings related to the office hours integration)
- Run build (to get the new paragraph type and related functionality)
- Use VPN (the API is restricted by IP)
- Go to edit any instruction (article)
- Add an "Office hours" paragraph
- Unselect any selected degree programme in the degree programme selector if one is selected
- Verify that all office hours are displayed
- Select a specific degree programme (in this case, the master's in data science should work)
- Verify that only the specific office hours (with the degree programme matching the selected degree programme) are displayed
- Turn off VPN and clear caches (`drush cr`)
- Reload the article
- Verify that no errors are displayed and no office hours are displayed
- Verify that the connection error has been logged in `var/log/messages`
